s21/p5/tester.py

```diff
@@ -0,0 +1,509 @@
+    except subprocess.TimeoutExpired:
+        print("Execution time for sample should be less than 90 seconds")
+        return 0
+    except Exception as e:
+        print(type(e), e)
+        
+    elapsed_time = time.time() - start_time
+    if elapsed_time > 20:
+        print("(-10) sample command can be optimized further. (<= 20secs), elapsed time:", elapsed_time)
+        print("Make sure you have an O(M+N) implementation")
+        points -= 10
+        
+    
+    # write on actual_json
+    actual_count = 0
+    for i, row in enumerate(zip_csv_iter(zout)):
+        actual_json[f"sample_{i}"] = row
+        actual_count += 1
+    
+    # dump in the middle for convenience
+    with open("actual.json", "w", encoding="utf-8") as f:
+        json.dump(actual_json, f, indent=2)
+    
+    # open expected
+    with open("expected.json") as f:
+        expected_json = json.load(f)
+    
+    expected_count = 0
+    for key in expected_json:
+        if "sample" in key:
+            expected_count += 1
+    
+    unit_points = points / expected_count
+    
+    # compare the number of rows
+    if expected_count != actual_count:
+        print(f"expected {expected_count} rows, found {actual_count} rows.")
+        points -= unit_points * abs(expected_count - actual_count)
+    
+    for i in range(expected_count):
+        try: # compare rows
+            actual_row = actual_json[f"sample_{i}"]
+            expected_row = expected_json[f"sample_{i}"]
+            if actual_row != expected_row:
+                print(f"row {i}")
+                print(f"expected row: {expected_row}")
+                print(f"found row: {actual}")
+        except IndexError as e:
+            print(f"missing row {i}: {expected_row}")
+        except Exception as e:
+            print(type(e), e)
+    
+    points = max(0, points)
+    return points
+@test(points=25)
+def world():
+    zname = "small.zip"
+    points = 25
+    
+    svg = "world_output.svg"
+    if os.path.exists(svg):
+        os.remove(svg)
+    run("world", zname, svg)
+    stats = svg_analyze(svg)
+    
+    if stats["paths"] < 270 or stats["paths"] > 300:
+        print("%s doesn't look like a world map" % svg)
+        return 0
+    
+    if stats["paths"] > 284:
+        print("ERROR: please remove Antartica")
+        points -= 10
+
+    if stats["colors"] < 3:
+        print("ERROR: use more different shades to represent traffic levels")
+        points -= 10
+
+    if stats["width"] < 450:
+        print("ERROR: make the plot wider")
+        points -= 5
+        
+    return points
+
+@test(points=25)
+def phone():
+    points = 25;
+    zname = "docs.zip"
+    
+    actual = run("phone", zname).strip().split("\n")
+    actual_json['phone'] = actual
+    
+    
+    # open and parse 'phone_expected.txt'
+    with open('expected.json', 'r') as f:
+        expected_json = json.load(f)
+        expected = set(expected_json['phone'])
+        
+    if len(actual) - len(set(actual)) > 0 :
```

Comment:
I was hoping for something like this that is less discrete:

weight = 20
points -= len(set(expected).difference(set(actual)) / len(expected) * weight

For missing numbers, you can print something like this:
sorted(set(expected).difference(set(actual))[:10]

Reply:
This part is about the requirement "no duplication" - so I think it would be better to use discrete deduction. For missing numbers, I followed yours now.